File: solutions/go/03/lis/main.go
package main

import "fmt"

func lis(a []int) int {
	n := len(a)
	dp := make([]int, n)
	for i := range dp {
		dp[i] = 1
	}
	ans := 0
	for i := 0; i < n; i++ {
		for j := 0; j < i; j++ {
			if a[j] < a[i] && dp[j]+1 > dp[i] {
				dp[i] = dp[j] + 1
			}
		}
		if dp[i] > ans {
			ans = dp[i]
		}
	}
	return ans
}

func main() {
	fmt.Println(lis([]int{3, 1, 4, 2, 5}))
}

File: solutions/python/03/lis/main.py
def lis(a: list) -> int:
    n = len(a)
    dp = [1] * n
    for i in range(n):
        for j in range(i):
            if a[j] < a[i]:
                dp[i] = max(dp[i], dp[j] + 1)
    return max(dp)


def main() -> None:
    print(lis([3, 1, 4, 2, 5]))


if __name__ == "__main__":
    main()
